lib/frame.c

```diff
@@ -84,19 +84,27 @@ int quicly_decode_ack_frame(const uint8_t **src, const uint8_t *end, quicly_ack_
     for (i = 0; i != frame->num_gaps; ++i) {
         if ((tmp = quicly_decodev(src, end)) == UINT64_MAX)
             goto Error;
-        frame->gaps[i] = tmp + 1;
-        if (frame->smallest_acknowledged < frame->gaps[i])
+        curr_gap = tmp + 1;
+        if (frame->smallest_acknowledged < curr_gap)
             goto Error;
-        frame->smallest_acknowledged -= frame->gaps[i];
         if ((tmp = quicly_decodev(src, end)) == UINT64_MAX)
             goto Error;
         tmp += 1;
-        if (frame->smallest_acknowledged < tmp)
-            goto Error;
-        frame->ack_block_lengths[i + 1] = tmp;
-        frame->smallest_acknowledged -= tmp;
+        if (i < QUICLY_ACK_MAX_GAPS) {
+            frame->smallest_acknowledged -= curr_gap;
+            if (frame->smallest_acknowledged < tmp)
+                goto Error;
+            frame->ack_block_lengths[i + 1] = tmp;
+            frame->smallest_acknowledged -= tmp;
+            frame->gaps[i] = curr_gap;
+        }
     }
 
+    /* the entire ACK-Ranges section was decoded, but ranges beyond
+       QUICLY_ACK_MAX_GAPS are skipped. therefore calibrate the value. */
+    if (frame->num_gaps > QUICLY_ACK_MAX_GAPS)
+        frame->num_gaps = QUICLY_ACK_MAX_GAPS;
```

Comment:
I think using `num_gaps` for two purposes is contributing to the complexity of the source code. Please see my comment above describing one way we might consider.